File: examples/environments/template_folders/test_harness.py
#!/usr/bin/env python3
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# Copyright (c) 2026 Ezequiel Alves. All rights reserved.

import sys
import os

sys.path.insert(0, os.path.abspath("work/python"))
import reg_map as reg_map

def main():
    print("[Python Harness] Testing template_folders Register Driver...")
    mem = {}
    def sim_read(addr):
        return mem.get(addr, 0)
    def sim_write(addr, val):
        mem[addr] = val

    block = reg_map.SAMPLE_BLOCKBlock(base_addr=0x0, read_fn=sim_read, write_fn=sim_write)
    reg = getattr(block, 'sample_reg')
    assert reg is not None
    print(f'Tested register {reg.name} at offset 0x{reg.offset:X}')
    if 'VAL' in reg.fields:
        block.write_field(reg, 'VAL', 1)
        assert block.read_field(reg, 'VAL') == 1
        print('Tested field VAL read/write')
    print("[Python Harness] All template_folders driver tests PASSED!")

if __name__ == "__main__":
    main()

File: examples/environments/template_folders/test_harness.c
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright (c) 2026 Ezequiel Alves. All rights reserved.
 */

#include <assert.h>
#include <stdio.h>
#include <stdint.h>
#include "reg_map.h"

int main(void) {
    printf("[C Harness] Testing template_folders Register Map Definitions...\n");
    printf("Checking SAMPLE_BLOCK_SAMPLE_REG_OFFSET = 0x%X\n", (unsigned int)SAMPLE_BLOCK_SAMPLE_REG_OFFSET);
    assert(SAMPLE_BLOCK_SAMPLE_REG_OFFSET >= 0);
    printf("Checking SAMPLE_BLOCK_SAMPLE_REG_VAL_MASK = 0x%X\n", (unsigned int)SAMPLE_BLOCK_SAMPLE_REG_VAL_MASK);
    assert(SAMPLE_BLOCK_SAMPLE_REG_VAL_MASK != 0);
    printf("Checking sizeof(sample_block_regs_t) = %zu\n", sizeof(sample_block_regs_t));
    assert(sizeof(sample_block_regs_t) > 0);
    printf("[C Harness] All template_folders assertions PASSED!\n");
    return 0;
}
